### ISSUE TITLE
Kafka Streams ObjectMapper modules

### ISSUE THREAD

**Issue by pasmikm** [spring-cloud/spring-cloud-stream-binder-kafka#812]
Hello, it seems that Kafka Streams binder in version 3.0.0 started using some custom ObjectMapper for deserializing data through Jackson and it lacks ParameterNamesModule. 
Previously we could use classes like

    @Data
    public class SomeData {
        @JsonProperty("OLD_STYLE_PROP")
        private final String oldStyleProp;
        @JsonProperty("BADNAMING")
        private final String badNaming;
    }
but currently KStreams only work with JsonCreator annotated constructor with each parameter annotated with JsonProperty.
Minimal application with embedded Kafka test is attached, if you switch to commented out versions of Spring Boot (2.1.7) and Spring Cloud Stream (Greenwich) it works, but it fails with 2.2.1 + Hoxton:
[object-mapper-test.zip](https://github.com/spring-cloud/spring-cloud-stream-binder-kafka/files/3951014/object-mapper-test.zip)

Is there some additional configuration necessary in 3.0.0 or could it be changed back to use the standard ObjectMapper from context (which includes modules)?

**Comment by sobychacko** [comment]
@pasmikm In order to restore the old behavior, you can do this:
`spring.cloud.stream.bindings.input.consumer.useNativeDecoding=false` in your test (and in the application as well).
This will force the deserialization done by the framework. 
With that said, it looks like there are some issues with native decoding in the binder and Spring Kafka `JsonSerde` (which is the default). We will try to see if we can troubleshoot this. While that's going on, can you set the decoding to be done by SCSt to get going?

**Comment by sobychacko** [comment]
@pasmikm Actually, if you change your data class as below, the default settings (`useNativeDecoding` set to `true`) work fine and the test passes.
```
@Data
@NoArgsConstructor
public class SomeData {
    
    @JsonProperty("OLD_STYLE_PROP")
    private String oldStyleProp;
    
    @JsonProperty("BADNAMING")
    private String badNaming;
}
```
Therefore, I don't see any issues from the binder perspective and suggest you go with this fix if you are staying with default `useNativeDecoding` settings.

**Comment by pasmikm** [comment]
Thanks for the reply, useNativeDecoding configuration change seems to work fine so far (as we would prefer not to change lots of data classes).

**Comment by tim-klug** [comment]
Hi @sobychacko,

just hit the same problem and I found the reason for my missing module here: (https://github.com/spring-projects/spring-kafka/blob/master/spring-kafka/src/main/java/org/springframework/kafka/support/JacksonUtils.java). I have the actual problem to serialize a `Money` field. I have tried the solution from above, but then I got a `JsonMappingException` with the message 'Currency Code may not be null'. So basically for some reason still the *included* `ObjectMapper` is used in between. I have a producer, which writes to a topic, and the message is serialized with the Zalando Money module. Then on the processor, the KStream uses a custom json serde which enhances the `ObjectMapper` with the Zalando module.
Now when the message is picked from the stream the deserializer runs into the problem, that the incoming json is not the one serialized by the producer. The json contains an object, which was serialized without the Zalando Money Module and cannot be deserialized back to a Java class with a type `Money` field.

Any idea on how to solve this?
Thx
Tim

**Comment by sobychacko** [comment]
@tim-klug Did you try a custom deserializer that uses the mentioned type from zalando? If the serializer is using the custom one, then you may want to use a custom one on the deserializing side as well.

**Comment by tim-klug** [comment]
Yes I tried this. I debugged the processor call and the custom deserializer was picked up. The interesting thing here is, that at that time the serialized byte array in the message looks is in a way serialized like if the module was missing.
So with checking on the producing side and on the consuming side, I can see, that on both sides the custom serializers are picked up. Somewhere in the middle a deserilization - serialization must occur therefor.
This behavior is showing up after upgrading from Spring Boot 2.1 to 2.3.9. So atm integration tests with an embedded kafka and messages containing `Money.class` objects are failing.

Is there some middle mechanism picking up the messages and deserialize and serialize them?

**Comment by sobychacko** [comment]
@tim-klug Interesting. Any chance where you can put this up in a small maven project  (with the test) where we can see the issue?

**Comment by tim-klug** [comment]
I'll do that and share it with you. Thanks for the quick response.

**Comment by tim-klug** [comment]
@sobychacko sorry for stressing. I found the issue ;-) 
The configuration in the `application.yml` was not correct. There are some StackOver answers out there where the key for the native encoding is not correct. The problem is, the because of the generic way there is linting giving a hint.
Anyway, I have created the demo project here: https://github.com/tim-klug/kafka-stream-money-deserialization

Interesting thing is, that I don't have to use the `useNativeDecoding: true` and only define a serde for the processor in `spring.cloud.stream.kafka.streams.binders: ...`

Any reasons why the `ObjectMapper` is not reused from the Spring context?

**Comment by sobychacko** [comment]
@tim-klug My apologies for the delayed response. I've been meaning to get back to you on this. Are you still having any issues are good at this point? `useNativeDecoding: true` is the default set by the binder. The only time you have to change this to `false` is when you want to use the framework provided message conversion. In any case, could you please let us know if you are having any more issues with this and whether we can close this?

**Comment by tim-klug** [comment]
It is working as expected.
Just curious why the ObjectMapper is not injected, so that it can be enhanced with modules.

**Comment by sobychacko** [comment]
What do you see when debugging? I mean, what is getting injected [here](https://github.com/tim-klug/kafka-stream-money-deserialization/blob/main/src/main/java/com/example/demo/configuration/KStreamConfigurations.java#L17)? 

You should get the standard `ObjectMapper` bean from [boot](https://github.com/spring-projects/spring-boot/blob/main/spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/jackson/JacksonAutoConfiguration.java#L100).

**Comment by tim-klug** [comment]
I can check, but I’m pretty sure, that the Mapper is the one I want. But then I will have to use the message converter. I saw that when we do not use the ObjectMapper, the standard implementation checks for four different modules and if present adds them. Why is the ObjectMapper here not taking from the context?
We just upgraded from Spring Boot 2.1 and before the update we didn’t had to add the serdes.
Just interested in the design decision here. Anything specific let to this?

**Comment by sobychacko** [comment]
Starting with 3.0.x versions of Kafka Streams binder, the default for deserialization is changed so that it now delegates to Kafka Streams. What that means is that instead of relying on message converters, the framework by default uses `Serde`s to do the deserialization and serialization. For the most part the binder can infer the right `Serde` type to use from your type signature. However, for anything advanced (as I believe in your case), where it needs to handle special cases, then the Serdes have to be provided. See this [section](https://docs.spring.io/spring-cloud-stream-binder-kafka/docs/3.1.2/reference/html/spring-cloud-stream-binder-kafka.html#_inbound_deserialization) from the ref docs for more details on that. There are a few ways you can provide the `Serde` - either in configs as you have it in the sample provided or by defining it as a `@Bean`. 

Bottom line is that if you have to use the `ObjectMapper`, then you have to wrap it inside your `Serde` implementation. You can also try disabling it altogether by setting `useNativeDecoding` to `false` and rely on the converters directly, but I won't recommend doing that as relying on `Serde` is a better mechanism in Kafka Streams. 

Another thing I wanted to point out to you after looking at your sample is that you might want to upgrade your app to use the functional model instead of using `EnableBinding` as the framework now largely favors the former and already deprecated the `EnableBinding` imperative style.

**Comment by tim-klug** [comment]
I saw the deprecated notice while doing the upgrade. I like the functional approach and we planned to migrate all processors in the next sprint.
Great work and thx for the support!

**Comment by sobychacko** [comment]
Please let us know if you run into further issues. I am closing this issue for the time being.
